mesonbuild/dependencies/dev.py

```diff
@@ -129,101 +133,96 @@ class LLVMDependency(ExternalDependency):
         # so strip the 'svn' off the end, since it will probably cuase problems
         # for users who want the patch version.
-        self.version = out.strip().rstrip('svn')
-
-        p, out = Popen_safe(
-            [self.llvmconfig, '--libs', '--ldflags'])[:2]
-        if p.returncode != 0:
-            raise DependencyException('Could not generate libs for LLVM.')
-        self.link_args = shlex.split(out)
-
-        p, out = Popen_safe([self.llvmconfig, '--cppflags'])[:2]
-        if p.returncode != 0:
-            raise DependencyException('Could not generate includedir for LLVM.')
-        cargs = mesonlib.OrderedSet(shlex.split(out))
+        self.version = self.version.rstrip('svn')
+
+        self.provided_modules = self.get_config_value(['--components'], 'modules')
+        modules = stringlistify(extract_as_list(kwargs, 'modules'))
+        self.check_components(modules)
+        opt_modules = stringlistify(extract_as_list(kwargs, 'optional_modules'))
+        self.check_components(opt_modules, required=False)
+
+        cargs = set(self.get_config_value(['--cppflags'], 'compile_args'))
         self.compile_args = list(cargs.difference(self.__cpp_blacklist))
 
-        p, out = Popen_safe([self.llvmconfig, '--components'])[:2]
-        if p.returncode != 0:
-            raise DependencyException('Could not generate modules for LLVM.')
-        self.modules = shlex.split(out)
+        if version_compare(self.version, '>= 3.9'):
+            self._set_new_link_args()
+        else:
+            self._set_old_link_args()
+        self.link_args = strip_system_libdirs(environment, self.link_args)
+
+    def _set_new_link_args(self):
+        """How to set linker args for LLVM versions >= 3.9"""
+        link_args = ['--link-static', '--system-libs'] if self.static else ['--link-shared']
+        self.link_args = self.get_config_value(
+            ['--libs', '--ldflags'] + link_args + list(self.required_modules),
+            'link_args')
+
+    def _set_old_link_args(self):
+        """Setting linker args for older versions of llvm.
+
+        Old versions of LLVM bring an extra level of insanity with them.
+        llvm-config will provide the correct arguments for static linking, but
+        not for shared-linnking, we have to figure those out ourselves, because
+        of course we do.
+        """
+        if self.static:
+            self.link_args = self.get_config_value(
+                ['--libs', '--ldflags', '--system-libs'] + list(self.required_modules),
+                'link_args')
+        else:
+            # llvm-config will provide arguments for static linking, so we get
+            # to figure out for ourselves what to link with. We'll do that by
+            # checking in the directory provided by --libdir for a library
+            # called libLLVM-<ver>.(so|dylib|dll)
+            libs = self.get_config_value(['--libdir'], 'link_args')[0]
```

Comment:
__[flake8]__
* [F841] local variable 'libs' is assigned to but never used

<!-- Comment by [SideCI](https://sideci.com) -->